src/detect.rs

```diff
@@ -18,3 +23,196 @@ pub enum DetectOutcome {
+        }
+    }
+
+    pub fn buildpack_toml(mut self, arg: BuildpackToml<BM>) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::BuildpackToml(arg));
+        self
+    }
+
+    pub fn buildpack_toml_from_string(
+        mut self,
+        string: impl AsRef<str>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::Str(string.as_ref().to_string()));
+        self
+    }
+
+    pub fn buildpack_toml_from_path(
+        mut self,
+        path: impl AsRef<Path>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.buildpack_toml = Some(BuildpackTomlArg::Path(path.as_ref().to_path_buf()));
+        self
+    }
+
+    /// Used for testing purposes to set a minimal viable context
+    ///
+    /// Warning: The directory passed in will be modified.
+    ///
+    ///```
+    /// use tempfile;
+    /// use libcnb::DetectContextBuilder;
+    ///
+    /// pub fn detect(context: libcnb::GenericDetectContext) {}
+    ///
+    /// let temp_dir = tempfile::tempdir().unwrap();
+    /// let context = DetectContextBuilder::new("heroku-20")
+    ///     .buildpack_toml_from_string(include_str!("../examples/example-02-ruby-sample/buildpack.toml"))
+    ///     .from_path(&temp_dir)
+    ///     .build::<std::io::Error>()
+    ///     .unwrap();
+    ///
+    /// detect(context);
+    ///```
+    pub fn from_path(mut self, dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
+        let dir = dir.as_ref();
+        let app_dir = dir.join("app");
+        let platform_dir = dir.join("platform");
+        let buildpack_dir = dir.join("buildpack");
+
+        for dir in [&app_dir, &platform_dir, &buildpack_dir] {
+            std::fs::create_dir_all(dir).unwrap();
+        }
+
+        if self.app_dir.is_none() {
+            self.app_dir = Some(app_dir);
+        }
+
+        if self.buildpack_dir.is_none() {
+            self.buildpack_dir = Some(buildpack_dir);
+        }
+
+        if self.platform.is_none() {
+            self.platform = Some(PlatformArg::Dir(platform_dir));
+        }
+        self
+    }
+
+    pub fn platform(mut self, platform: P) -> DetectContextBuilder<P, BM> {
+        self.platform = Some(PlatformArg::Plat(platform));
+        self
+    }
+
+    pub fn platform_from_path(
+        mut self,
+        platform_dir: impl AsRef<Path>,
+    ) -> DetectContextBuilder<P, BM> {
+        self.platform = Some(PlatformArg::Dir(platform_dir.as_ref().to_path_buf()));
+        self
+    }
+
+    pub fn app_dir(mut self, app_dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
+        self.app_dir = Some(app_dir.as_ref().to_path_buf());
+        self
+    }
+
+    pub fn buildpack_dir(mut self, buildpack_dir: impl AsRef<Path>) -> DetectContextBuilder<P, BM> {
+        self.buildpack_dir = Some(buildpack_dir.as_ref().to_path_buf());
+        self
+    }
+
+    pub fn build<E: Display + Debug>(self) -> crate::Result<DetectContext<P, BM>, E> {
```

Comment:
I don't think this should return a `libcnb::Result` but something specific for this builder. We use `Other` liberally for builder specific errors here. Bottom line, our builder errors are just `String`s instead of proper values.

Reply:
Errors are an extreme weak spot i'll defer to @hone here.

Reply:
@Malax I refactored some stuff in #97, there's still many error cases covered by `libcnb::Result`, we could just make our own error type if there are other cases we want to cover.